libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/TelemetryEventUserFeedback.kt

```diff
@@ -0,0 +1,89 @@
+package com.mapbox.navigation.core.telemetry
+
+import android.location.Location
+import com.google.gson.Gson
+
+internal enum class FeedbackType(val feedbackType: String) {
+    FEEDBACK_TYPE_ACCIDENT("accident"),
+    FEEDBACK_TYPE_HAZARD("hazard"),
+    FEEDBACK_TYPE_ROAD_CLOSED("road_closed"),
+    FEEDBACK_TYPE_NOT_ALLOWED("not_allowed"),
+    FEEDBACK_TYPE_ROUTING_ERROR("routing_error"),
+    FEEDBACK_TYPE_MISSING_ROAD("missing_road"),
+    FEEDBACK_TYPE_MISSING_EXIT("missing_exit"),
+    FEEDBACK_TYPE_CONFUSING_INSTRUCTION("confusing_instruction"),
+    FEEDBACK_TYPE_INACCURATE_GPS("inaccurate_gps"),
+    FEEDBACK_SOURCE_REROUTE("reroute"),
+    FEEDBACK_SOURCE_UI("user")
+}
+
+internal enum class FeedbackSource(val feedbackSource: String) {
+    FEEDBACK_SOURCE_USER("user"),
+    FEEDBACK_SOURCE_REROUTE("reroute"),
+    FEEDBACK_SOURCE_UNKNOWN("unkknown")
+}
+
+internal data class TelemetryStep(
+    var upcomingType: String = "",
+    var upcomingModifier: String = "",
+    var upcomingName: String = "",
+    var previousType: String = "",
+    var previousModifier: String = "",
+    var previousName: String = "",
+    var distance: Int = 0,
+    var duration: Int = 0,
+    var distanceRemaining: Int = 0,
+    var durationRemaining: Int = 0
+)
+
+/**
+ * This class contains required data for a feedback event. The caller posting a feedback event
+ * does not need to provide it.
+ */
+internal data class TelemetryUserFeedbackWrapper(
+    val userFeedback: TelemetryEventFeedback,
+    val userId: String,
+    val audio: String,
+    val locationsBefore: Array<Location>,
+    val locationsAfter: Array<Location>,
+    val feedbackId: String,
+    val screenshot: String,
+    val step: TelemetryStep
+) : TelemetryEventInterface, MetricEvent {
+
+    override val metricName = NavigationMetrics.FEEDBACK
+
+    override fun toJson(gson: Gson) =
+            gson.toJson(this)
+
+    // IDE Generated code
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as TelemetryUserFeedbackWrapper
+
+        if (userFeedback != other.userFeedback) return false
+        if (userId != other.userId) return false
+        if (audio != other.audio) return false
+        if (!locationsBefore.contentEquals(other.locationsBefore)) return false
+        if (!locationsAfter.contentEquals(other.locationsAfter)) return false
+        if (feedbackId != other.feedbackId) return false
+        if (screenshot != other.screenshot) return false
+        if (step != other.step) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
```

Comment:
don't need it for data class

Reply:
True in general, but this data class contains Arrays,  so the IDE is suggesting I implement these. 

> Good progress @olegzil. Adding to all other comments, I'd love to see an initial test structure implementation. The concepts you're introducing might allow us to collect metrics from any thread (another question is, do we need it?), but it'd be great if the architecture was designed in a testable manner from the get-go.

I am working on a set of Unit tests that should test this code extensively. Regarding the need for thread safety. Two answers regarding this subject. In this particular case the Telemetry class exposes one public method (not counting initialized()) and subscribes to two interfaces that have a combined three callbacks. 
**postTelemetryEvent()** -- a method that can be called from any thread. This method cannot block the caller. It is used to report telemetry events from any part of the SDK, running on any thread.
The call backs are:
**RouteProgressObserver** and **LocationEngineCallback**. These two are usually called from the Main thread.  Both the public method and the callbacks operate on the same data: **locationBuffer** and **telemetryEventsQueue**.  As you can see, since object state can potentially be modified from multiple threads, access to the state should be serialized.

On a more general note, the SDK has multiple Singletons, each maintains a state that can be accessed via the singletons public methods. We should consider making these singletons thread safe. Finally, do we want the SDK to be reactive? Making it reactive requires that no public method blocks the caller in addition to guaranteeing thread sefety.

Reply:
👍